crux-http-server/src/crux/http_server.clj

```diff
@@ -380,14 +368,167 @@
       ((resolve 'crux.sparql.protocol/sparql-query) crux-node request)
       nil)))
 
+(defn link-all-entities
+  [crux-node vt tt path result]
+  (let [resolved-links (fn recur-on-result [result links]
+                         (if (and (c/valid-id? result)
+                                  (api/entity (api/db crux-node vt tt) result))
+                           (let [query-params (cond-> "?"
+                                                vt (str "valid-time=" vt "&")
+                                                tt (str "transact-time=" tt))]
+                             (assoc links result (str path "/" result query-params)))
+                           (cond
+                             (map? result) (reduce-kv #(merge %1 (recur-on-result %3 %1)) links result)
+                             (sequential? result) (reduce #(merge %1 (recur-on-result %2 %1)) links result)
+                             :else links)))]
+    (resolved-links result {})))
+
+(defn- entity->html [links edn]
+  (if-let [href (get links edn)]
+    [:a {:href href} (str edn)]
+    (cond
+      (map? edn) (into [:dl]
+                       (mapcat
+                        (fn [[k v]]
+                          [[:dt (entity->html links k)]
+                           [:dd (entity->html links v)]])
+                        edn))
+      (sequential? edn) (into [:ol] (map (fn [v] [:li (entity->html links v)]) edn))
+      (set? edn) (into [:ul] (map (fn [v] [:li (entity->html links v)]) edn))
+      :else (str edn))))
+
+(defn- entity-state [^ICruxAPI crux-node request]
+  (let [[_ encoded-eid] (re-find #"^/_entity/(.+)$" (req/path-info request))]
+    (let [eid (c/id-edn-reader encoded-eid)
+          query-params (:query-params request)
+          valid-time (some-> (get query-params "valid-time")
+                             (instant/read-instant-date))
+          transaction-time (some-> (get query-params "transaction-time")
+                                   (instant/read-instant-date))
+          db (db-for-request crux-node {:valid-time valid-time
+                                        :transact-time transaction-time})
+          entity-map (api/entity db eid)]
+      {:status (if (some? entity-map) 200 404)
+       :body (when entity-map
+               (if (= (get-in request [:muuntaja/response :format]) "text/html")
+                 (let [linked-entities (link-all-entities crux-node valid-time transaction-time "/_entity" entity-map)]
+                   (html5 (entity->html linked-entities entity-map)))
+                 entity-map))})))
+
+(defn- coerce-param
+  ([param]
+   (coerce-param param nil))
+  ([param wrap?]
+   (let [formatted-param
+         (cond
+           (vector? param) (str "[" (reduce str param) "]")
+           wrap? (str "[" param "]")
+           :else param)]
+     (c/read-edn-string-with-readers formatted-param))))
```

Comment:
seems strange that we're giving the user the choice of EDN/JSON/etc response types, but mandating EDN query params - haven't come across this before. thinking about it, whenever I've had something sufficiently complicated to warrant EDN (i.e. more than numbers, bools, dates, etc) in query params, I've usually moved it to a request body instead, at which point you've got request-body coercion using the `content-type` header - but this goes against the 'hackable query URL' principle.

again, likely one for @malcolmsparks's opinion

Reply:
from @malcolmsparks: fine to keep EDN-formatted QPs for now, especially as we have the raw `q=` param too